ibis/backends/datafusion/client.py

```diff
@@ -0,0 +1,204 @@
+from typing import Dict, List, Mapping
+
+import pyarrow as pa
+import datafusion as df
+
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.schema as sch
+import ibis.expr.types as ir
+from ibis.backends.base import Client, Database
+
+
+# TODO(kszucs): support nested and parametric types
+# consolidate with the logic from the parquet backend
+
+_to_ibis_dtypes = {
+    pa.int8(): dt.Int8,
+    pa.int16(): dt.Int16,
+    pa.int32(): dt.Int32,
+    pa.int64(): dt.Int64,
+    pa.uint8(): dt.UInt8,
+    pa.uint16(): dt.UInt16,
+    pa.uint32(): dt.UInt32,
+    pa.uint64(): dt.UInt64,
+    pa.float16(): dt.Float16,
+    pa.float32(): dt.Float32,
+    pa.float64(): dt.Float64,
+    pa.string(): dt.String,
+    pa.binary(): dt.Binary,
+    pa.bool_(): dt.Boolean,
+    pa.timestamp('ns'): dt.Timestamp,
+    pa.timestamp('ms'): dt.Timestamp
+}
+
+_to_pyarrow_types = {
+    dt.Int8: pa.int8(),
+    dt.Int16: pa.int16(),
+    dt.Int32: pa.int32(),
+    dt.Int64: pa.int64(),
+    dt.UInt8: pa.uint8(),
+    dt.UInt16: pa.uint16(),
+    dt.UInt32: pa.uint32(),
+    dt.UInt64: pa.uint64(),
+    dt.Float16: pa.float16(),
+    dt.Float32: pa.float32(),
+    dt.Float64: pa.float64(),
+    dt.String: pa.string(),
+    dt.Binary: pa.binary(),
+    dt.Boolean: pa.bool_(),
+    dt.Timestamp: pa.timestamp('ns'),
+}
+
+# TODO(kszucs): the following conversions are really rudimentary
+# we should have a pyarrow backend which would be responsible
+# for conversions between ibis types to pyarrow types
+
+def to_pyarrow_type(dtype):
+    #str(dtype))
+    try:
+        return _to_pyarrow_types[dtype.__class__]
+    except KeyError as e:
+        if isinstance(dtype, dt.Array):
+            return pa.list_(to_pyarrow_type(dtype.value_type))
+        raise e
+
+
+@dt.dtype.register(pa.DataType)
+def from_pyarrow_type(arrow_type, nullable=True):
+    return _to_ibis_dtypes[arrow_type](nullable=nullable)
+
+
+@sch.infer.register(pa.Schema)
+def infer_pyarrow_schema(schema):
+    fields = [(f.name, dt.dtype(f.type, nullable=f.nullable)) for f in schema]
+    return sch.schema(fields)
+
+
+class DataFusionTable(ops.DatabaseTable):
+    pass
+
+
+class DataFusionDatabase(ops.DatabaseTable):
+    pass
+
+
+class DataFusionClient(Client):
+    def __init__(self, backend, config):
+        self.backend = backend
+
+        if isinstance(config, df.ExecutionContext):
+            self.context = config
+        else:
+            self.context = df.ExecutionContext()
+            for name, path in config.items():
+                if path.endswith('.csv'):
+                    self.register_csv(name, path)
+                elif path.endswith('.parquet'):
+                    self.register_parquet(name, path)
+                else:
+                    raise ValueError('Wrong format')
```

Comment:
Maybe something a little more informative like `"The DataFusion backend only supports CSV files with the extension .csv and Parquet files with the .parquet extension"`